automod/countstore/countstore.go

```diff
@@ -0,0 +1,54 @@
+package countstore
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"time"
+)
+
+const (
+	PeriodTotal = "total"
+	PeriodDay   = "day"
+	PeriodHour  = "hour"
+)
+
+// CountStore is an interface for storing incrementing event counts, bucketed into periods.
+// It is implemented by MemCountStore and by RedisCountStore.
+//
+// Period bucketing works on the basis of the current date (as determined mid-call).
+// See the `Period*` consts for the available period types.
+//
+// The "GetCount" and "Increment" methods perform actual counting.
+// The "*Distinct" methods have a different behavior:
+// "IncrementDistinct" marks a value as seen at least once,
+// and "GetCountDistinct" asks how _many_ values have been seen at least once.
+//
+// The exact implementation and precision of the "*Distinct" methods may vary:
+// in the MemCountStore implementation, it is precise (it's based on large maps);
+// in the RedisCountStore implementation, it uses the Redis "pfcount" feature,
+// which is based on a HyperLogLog datastructure which has probablistic properties
+// (see https://redis.io/commands/pfcount/ ).
```

Comment:
this is great, thanks!

we could also note here that `MemCountStore` will grow unbounded in memory over time. this will always be the case for `PeriodTotal` counts, but in particular for distinct maps, and that the hour/day entries are never cleared. which is an intentional simplification for now.

I can take this as a task for me to add

Reply:
added docs on this, good idea